mirgecom/euler.py

```diff
@@ -20,16 +22,214 @@
+#    dt_geometric_factor,
+#    dt_non_geometric_factor,
+# )
+
+__doc__ = """
+.. autofunction:: inviscid_operator
+"""
+
+
+#
+# Euler flow eqns:
+# d_t(q) + nabla .dot. f = 0 (no sources atm)
+# state vector q: [rho rhoE rhoV]
+# flux tensor f: [rhoV (rhoE + p)V (rhoV.x.V + p*I)]
+#
+
+
+def _interior_trace_pair(discr, vec):
+    i = discr.interp("vol", "int_faces", vec)
+    e = with_object_array_or_scalar(
+        lambda el: discr.opposite_face_connection()(el.queue, el), i
+    )
+    return TracePair("int_faces", i, e)
+
+
+def _inviscid_flux(discr, q, eos=IdealSingleGas()):
+
+    ndim = discr.dim
+
+    # q = [ rho rhoE rhoV ]
+    mass = q[0]
+    energy = q[1]
+    mom = q[2:]
+
+    p = eos.pressure(q)
+
+    # Fluxes:
+    # [ rhoV (rhoE + p)V (rhoV.x.V + p*I) ]
+    momflux = make_obj_array(
+        [
+            (mom[i] * mom[j] / mass + (p if i == j else 0))
+            for i in range(ndim)
+            for j in range(ndim)
+        ]
+    )
+    massflux = mom * make_obj_array([1.0])
+    energyflux = mom * make_obj_array([(energy + p) / mass])
+
+    flux = flat_obj_array(massflux, energyflux, momflux,)
+
+    return flux
+
+
+def _get_wavespeed(w, eos=IdealSingleGas()):
+
+    mass = w[0]
+    mom = w[2:]
+
+    v = mom * make_obj_array([1.0 / mass])
+
+    sos = eos.sound_speed(w)
+    wavespeed = clmath.sqrt(np.dot(v, v)) + sos
+    return wavespeed
+
+
+def _facial_flux(discr, w_tpair, eos=IdealSingleGas()):
+
+    dim = discr.dim
+
+    mass = w_tpair[0]
+    energy = w_tpair[1]
+    mom = w_tpair[2:]
+
+    normal = with_queue(mass.int.queue, discr.normal(w_tpair.dd))
+
+    # Get inviscid fluxes [rhoV (rhoE + p)V (rhoV.x.V + p*I) ]
+    qint = flat_obj_array(mass.int, energy.int, mom.int)
+    qext = flat_obj_array(mass.ext, energy.ext, mom.ext)
+
+    # - Figure out how to manage grudge branch dependencies
+    #    qjump = flat_obj_array(rho.jump, rhoE.jump, rhoV.jump)
+    qjump = qext - qint
+    flux_int = _inviscid_flux(discr, qint, eos)
+    flux_ext = _inviscid_flux(discr, qext, eos)
+
+    # Lax/Friedrichs/Rusonov after JSH/TW Nodal DG Methods, p. 209
+    flux_jump = (flux_int + flux_ext) * make_obj_array([0.5])
+
+    # wavespeeds = [ wavespeed_int, wavespeed_ext ]
+    wavespeeds = [_get_wavespeed(qint), _get_wavespeed(qext)]
+
+    lam = clarray.maximum(wavespeeds[0], wavespeeds[1])
+    lfr = qjump * make_obj_array([0.5 * lam])
+
+    # Surface fluxes should be inviscid flux .dot. normal
+    # rhoV .dot. normal
+    # (rhoE + p)V  .dot. normal
+    # (rhoV.x.V)_1 .dot. normal
+    # (rhoV.x.V)_2 .dot. normal
+    nflux = flat_obj_array(
+        [
+            np.dot(flux_jump[(i * dim): ((i + 1) * dim)], normal)
+            for i in range(dim + 2)
+        ]
+    )
+
+    # add Lax/Friedrichs term
+    flux_weak = nflux + lfr
+
+    return discr.interp(w_tpair.dd, "all_faces", flux_weak)
+
+
+def inviscid_operator(
+    discr, w, t=0.0, eos=IdealSingleGas(), boundaries={BTAG_ALL: DummyBoundary()},
```

Comment:
Not a fan of defaulting the boundaries. It's cute at first, until it isn't. Also, mutable defaults are asking for trouble. (`dict`s are mutable.)